[pslql] ParseException when parsing ELSIF

**Affects PMD Version:** 6.12.0

**Description:**
I’m trying to test PMD with PLSQL and I’m getting an error when parsing  <IF> <THEN> <ELSIF> <ELSE> <ENDIF>

When I use IF whitout ELSIF it works fine but when I use it i got an exception.

<details>
  <summary>Exception stacktrace</summary>
  <p>

```
net.sourceforge.pmd.PMDException: Error while parsing C:\sapatin\pmd-bin-6.12.0\update_planned_hrs.sql
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:99)
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:51)
        at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:78)
        at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:24)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: net.sourceforge.pmd.lang.plsql.ast.ParseException: Encountered " <UNSIGNED_NUMERIC_LITERAL> "2 "" at line 15, column 11.
...
```

</p>
</details>

**Code Sample demonstrating the issue:**

```
CREATE OR REPLACE PACKAGE BODY update_planned_hrs
IS

FUNCTION existing_planned (p_emp_id IN NUMBER, p_project_id IN NUMBER) RETURN NUMBER

IS

existing_hours NUMBER(4);

BEGIN
   IF  1 = 1   THEN
    
     null;
     
   ELSIF 2 = 2  THEN
     null;
   ELSE
     null;
     
   END IF;


   RETURN (existing_hours);

   EXCEPTION
          WHEN OTHERS THEN NULL;

   END existing_planned;
   

END update_planned_hrs;
/
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Running from command line with JDK 11.
`pmd -d C:\pmd-bin-6.12.0\test -l plsql -R rulesets/plsql/TomKytesDespair.xml -f xml`

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -63,6 +63,7 @@ The designer will still be shipped with PMD's binaries.
     *   [#1674](https://github.com/pmd/pmd/issues/1674): \[java] documentation of CommentDefaultAccessModifier is wrong
 *   plsql
     *   [#1731](https://github.com/pmd/pmd/issues/1731): \[pslql] ParseException when parsing ELSIF
+    *   [#1733](https://github.com/pmd/pmd/issues/1733): \[plsql] % not supported in "TestSearch%notfound"
 
 ### API Changes
 

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -61,6 +61,8 @@ The designer will still be shipped with PMD's binaries.
     *   [#1701](https://github.com/pmd/pmd/issues/1701): \[java] UseTryWithResources does not handle multiple argument close methods
 *   java-codestyle
     *   [#1674](https://github.com/pmd/pmd/issues/1674): \[java] documentation of CommentDefaultAccessModifier is wrong
+*   plsql
+    *   [#1731](https://github.com/pmd/pmd/issues/1731): \[pslql] ParseException when parsing ELSIF
 
 ### API Changes
 

--- a/pmd-plsql/etc/grammar/PldocAST.jjt
+++ b/pmd-plsql/etc/grammar/PldocAST.jjt
@@ -1509,6 +1509,14 @@ ASTSimpleExpression SimpleExpression() :
         TableName() { sb.append(token.image); } "." { sb.append(token.image); }
         Column() { sb.append(token.image); }
  |
+    // Named Cursor: https://docs.oracle.com/en/database/oracle/oracle-database/18/lnpls/named-cursor-attribute.html#GUID-CD8D8415-FF19-4D81-99BA-7825FD40CC96
+    LOOKAHEAD(3)
+        Column() { sb.append(token.image); } "%" ( LOOKAHEAD({getToken(1).image.equalsIgnoreCase("isopen")}) KEYWORD("ISOPEN")
+                     | LOOKAHEAD({getToken(1).image.equalsIgnoreCase("found")}) KEYWORD("FOUND")
+                     | LOOKAHEAD({getToken(1).image.equalsIgnoreCase("notfound")}) KEYWORD("NOTFOUND")
+                     | LOOKAHEAD({getToken(1).image.equalsIgnoreCase("rowcount")}) KEYWORD("ROWCOUNT")
+                     ) { sb.append('%').append(token.image); }
+ |
     LOOKAHEAD(2)
         Column() { sb.append(token.image); }
  )

--- a/pmd-plsql/etc/grammar/PldocAST.jjt
+++ b/pmd-plsql/etc/grammar/PldocAST.jjt
@@ -2085,7 +2085,7 @@ ASTElseClause ElseClause()  :
 ASTElsifClause ElsifClause()  :
 {}
 {
-      <ELSIF>  Expression() <THEN> (Statement())+
+      <ELSIF>  Expression() <THEN> (LOOKAHEAD(1, {!(getToken(1).kind == ELSIF)}) Statement())+
       { return jjtThis ; }
 }
 
@@ -2280,7 +2280,7 @@ ASTWhileStatement WhileStatement()  :
 ASTIfStatement IfStatement()  :
 {}
 {
-	<IF> Expression() <THEN> (Statement())+
+	<IF> Expression() <THEN> (LOOKAHEAD(1, {!(getToken(1).kind == ELSIF)}) Statement())+
 	( ElsifClause() {jjtThis.setHasElse();} )*
 	[ ElseClause() {jjtThis.setHasElse();} ]
 	<END> <IF>
@@ -6095,7 +6095,6 @@ ASTID ID(): {}
 		//| <DISTINCT>  //RESERVED WORD
 		| <DO>
 		| <DROP>  //RESERVED WORD
-		//| <ELSE> //SYNTAX //RESERVED WORD
 		| <ELSIF> //SYNTAX
 
 		//| <END> |<CURRENT_USER>
@@ -6339,7 +6338,6 @@ ASTQualifiedID QualifiedID(): {}
 		//<DISTINCT>
 		| <DO>
 		//<DROP>
-		//<ELSE>
 		| <ELSIF>
 		//<END>
 		//<EXCEPTION>

--- a/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/CursorAttributesTest.java
+++ b/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/CursorAttributesTest.java
@@ -1,0 +1,22 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.plsql.ast;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import net.sourceforge.pmd.lang.plsql.AbstractPLSQLParserTst;
+
+public class CursorAttributesTest extends AbstractPLSQLParserTst {
+
+    @Test
+    public void parseCursorWithAttribute() {
+        String code = loadTestResource("CursorAttributes.pls");
+        ASTInput input = parsePLSQL(code);
+        ASTExpression exp = input.getFirstDescendantOfType(ASTIfStatement.class).getFirstChildOfType(ASTExpression.class);
+        Assert.assertEquals("TestSearch%notfound", exp.getImage());
+    }
+
+}

--- a/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/IfStatementTest.java
+++ b/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/IfStatementTest.java
@@ -1,0 +1,20 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.plsql.ast;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import net.sourceforge.pmd.lang.plsql.AbstractPLSQLParserTst;
+
+public class IfStatementTest extends AbstractPLSQLParserTst {
+
+    @Test
+    public void parseIfWithElseIf() throws Exception {
+        String code = "BEGIN\nIF 1 = 1 THEN null;\nELSIF (2 = 2) THEN null;\nELSE null;\nEND IF;\nEND;\n/\n";
+        ASTInput input = parsePLSQL(code);
+        Assert.assertNotNull(input);
+    }
+}

--- a/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/CursorAttributes.pls
+++ b/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/CursorAttributes.pls
@@ -1,0 +1,32 @@
+CREATE OR REPLACE PACKAGE lpe_test is
+
+procedure testif;
+
+end lpe_test;
+/
+
+CREATE OR REPLACE PACKAGE BODY lpe_test is
+
+
+procedure testif
+is
+cursor TestSearch
+is
+select 1 from dual;
+
+TestRow TestSearch%rowtype;
+
+begin
+    open TestSearch;
+    fetch TestSearch into TestRow;
+
+    if TestSearch%notfound then
+        DBMS_OUTPUT.PUT_LINE('Not found');
+    else
+        DBMS_OUTPUT.PUT_LINE('Found');
+    end if;
+    close TestSearch;
+end testif;
+
+end lpe_test;
+/

